### PR TITLE
Implemented custom data stores for general and admin settings

### LINKED ISSUE #4461: PreferenceFragment implementations should use new preferences setup
The `PreferenceFragment` implementations should use the new preferences setup introduced in ##4459 rather than just the default `SharedPreferences` bucket. This most likely means exposing the names/ids for the preferences buckets from a provider.

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #4461

#### What has been done to verify that this works as intended?
I reviewed implemented changes, ran automated tests and played a bit with the app.

#### Why is this the best possible solution? Were any other approaches considered?
This is what's recommended in the docs https://developer.android.com/guide/topics/ui/settings/use-saved-values#custom-data-store

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
It doesn't require testing for now (since we decided to merge pr related to multi-project not to block us) but generally, general and admin settings should be tested.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/preferences/stores/AdminPreferencesDataStore.kt [hunk 1]
```diff
@@ -0,0 +1,54 @@
+package org.odk.collect.android.preferences.stores
+
+import androidx.preference.PreferenceDataStore
+import org.odk.collect.android.preferences.PreferencesDataSourceProvider
+
+class AdminPreferencesDataStore(val preferencesDataSourceProvider: PreferencesDataSourceProvider) : PreferenceDataStore() {
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/preferences/stores/AdminPreferencesDataStore.kt]: It feels we could just have a `PreferencesDataSourceDataStore` that takes a `PreferenceDataSource` in its constructor rather than specific implementations for admin and general. Is there a blocker for doing that?

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/preferences/stores/AdminPreferencesDataStore.kt]: I'm also thinking we might want to rename our `PreferenceDataSource` interface to avoid any confusion. I guess `PreferenceDataStore` will only be used in the context of `PreferenceFragment` though.

File: collect_app/src/main/java/org/odk/collect/android/preferences/BaseAdminPreferencesFragment.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import javax.inject.Inject;
+
+public abstract class BaseAdminPreferencesFragment extends BasePreferenceFragment {
+    @Inject
+    AdminPreferencesDataStore adminPreferencesDataStore;
+
+    @Override
+    public void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        getPreferenceManager().setPreferenceDataStore(adminPreferencesDataStore);
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/preferences/BaseAdminPreferencesFragment.java]: If we did have one `PreferenceDataStore` implementation we could just inject the `PreferenceDataSourceProvider` here instead of the specific implementation.

File: collect_app/src/test/java/org/odk/collect/android/preferences/stores/AdminPreferencesDataStoreTest.kt [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    @Before
+    fun setup() {
+        preferencesDataSourceProvider = mock(PreferencesDataSourceProvider::class.java)
+        adminPreferences = mock(PreferencesDataSource::class.java)
+        `when`(preferencesDataSourceProvider.getAdminPreferences()).thenReturn(adminPreferences)
+        adminPreferencesDataStore = AdminPreferencesDataStore(preferencesDataSourceProvider)
+    }
+
+    @Test
+    fun `When putString() called, should save() be called on adminPreferences`() {
```
Comment (seadowg) [inline collect_app/src/test/java/org/odk/collect/android/preferences/stores/AdminPreferencesDataStoreTest.kt]: I think for test names like this you can omit "when". I'd usually try and save it (or "given") for if I'm describing a condition. So for instance I'd name this test `putString saves string in adminPreferences`.  If the method exploded when you passed it a `null` key I'd write a test like `putString, when key is null, throws exception`. I'm also not including "should" here as it's not really needed. 

This is massively subjective so interested in your thoughts here! I think we have a nice opportunity to discuss different styles and then add some thoughts to our `CODE-GUIDELINES.md`.

File: collect_app/src/test/java/org/odk/collect/android/preferences/stores/AdminPreferencesDataStoreTest.kt [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        adminPreferencesDataStore.putString("Key", "value")
+        verify(adminPreferences, times(1)).save("Key", "value")
```
Comment (seadowg) [inline collect_app/src/test/java/org/odk/collect/android/preferences/stores/AdminPreferencesDataStoreTest.kt]: Don't think you need the `times` in these tests.

File: collect_app/src/test/java/org/odk/collect/android/preferences/stores/AdminPreferencesDataStoreTest.kt [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    @Test
+    fun `When putStringSet() called, should save() be called on adminPreferences`() {
+        adminPreferencesDataStore.putStringSet("Key", setOf("x", "y", "z"))
+        verify(adminPreferences, times(1)).save("Key", setOf("x", "y", "z"))
+    }
+
+    @Test
+    fun `When getString() called, should getString() be called on adminPreferences`() {
+        adminPreferencesDataStore.getString("Key", "value")
+        verify(adminPreferences, times(1)).getString("Key")
```
Comment (seadowg) [inline collect_app/src/test/java/org/odk/collect/android/preferences/stores/AdminPreferencesDataStoreTest.kt]: This doesn't actually verify that we are returning the right thing from `AdminPreferencesDataStore#getString`. We either use `when` to set up the return values and assert they come back or just create a fake (probably backed by a map) `PreferenceDataSource` implementation that we can add values to.  I think for simple cases like this the fake might be easier.

**Review by seadowg — CHANGES_REQUESTED** [review]
A lot of moving around here, but it looks like a great set of changes! I've left comments on a couple of things I think we should change and on some tests I think it'd be good to discuss before moving on.

File: collect_app/src/test/java/org/odk/collect/android/preferences/stores/AdminPreferencesDataStoreTest.kt (hunk 4 above)
Comment (grzesiek2010) [reply collect_app/src/test/java/org/odk/collect/android/preferences/stores/AdminPreferencesDataStoreTest.kt]: Right, Fixed.

File: collect_app/src/test/java/org/odk/collect/android/preferences/stores/AdminPreferencesDataStoreTest.kt (hunk 3 above)
Comment (grzesiek2010) [reply collect_app/src/test/java/org/odk/collect/android/preferences/stores/AdminPreferencesDataStoreTest.kt]: I agree, Fixed.

File: collect_app/src/main/java/org/odk/collect/android/preferences/BaseAdminPreferencesFragment.java (hunk 2 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/preferences/BaseAdminPreferencesFragment.java]: It's fixed now I inject `PreferencesDataStore`

File: collect_app/src/main/java/org/odk/collect/android/preferences/stores/AdminPreferencesDataStore.kt (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/preferences/stores/AdminPreferencesDataStore.kt]: Yeah I was thinking about this but somehow ended up with two implementation I can't recall the reason but now it's seems doable, what I did and now I have one PreferenceDataStore.

File: collect_app/src/test/java/org/odk/collect/android/preferences/stores/AdminPreferencesDataStoreTest.kt (hunk 5 above)
Comment (grzesiek2010) [reply collect_app/src/test/java/org/odk/collect/android/preferences/stores/AdminPreferencesDataStoreTest.kt]: Ok I improved that.

File: collect_app/src/test/java/org/odk/collect/android/preferences/source/PreferencesDataStoreTest.kt [hunk 6]
```diff
@@ -12,78 +15,101 @@ class PreferencesDataStoreTest {
     @Before
     fun setup() {
         preferencesDataSource = mock(PreferencesDataSource::class.java)
+        `when`(preferencesDataSource.getString(KEY)).thenReturn(STRING_ANSWER)
```
Comment (seadowg) [inline collect_app/src/test/java/org/odk/collect/android/preferences/source/PreferencesDataStoreTest.kt]: This is a bit nitpicky, but personally I'd try and avoid doing this kind of setup in a `Before`. I find when I'm reading a test it's nice to be able to see as much as possible of the state setup without scrolling or moving. I also find having constants set up like this harder to read than just having raw values or just local variables (all caps is probably the culprit). 

Another point is that I don't think we need to verify calls to `getString` etc as we're checking that that's happening through our state setup - we're testing the behaviour rather than the implementation.

I'd maybe suggest something like this for the `getString` read test:

```
@Test
fun `getString() reads string from preferences`() {
    `when`(preferencesDataSource.getString("key")).thenReturn("value")
    assertThat(preferencesDataStore.getString("key, "value"), `is`("value"))
}
```

Again, there are all just my opinions so lets discuss if you disagree with any of them!

File: collect_app/src/main/java/org/odk/collect/android/preferences/source/PreferencesDataStore.kt [hunk 7]
```diff
@@ -0,0 +1,53 @@
+package org.odk.collect.android.preferences.source
+
+import androidx.preference.PreferenceDataStore
+
+class PreferencesDataStore(val preferencesDataSource: PreferencesDataSource) : PreferenceDataStore() {
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/preferences/source/PreferencesDataStore.kt]: Ach my feedback has definitely pushed us towards some awkward naming here. I think this class should be named `<where data comes from>PreferenceDataStore` which would unfortunately mean calling this `PreferencesDataSourcePreferenceDataStore`. Yuck.

Maybe we should try and move away from using "preferences" for our own settings interfaces as Android seems to have done a good job of using every possible name using that word. Could we call our `PreferencesDataSource` something like `SettingsStore`, `ConfigStore`, `SettingsSource` etc?

Going to tag @lognaturel in as well because naming is hard.

**Review by seadowg — CHANGES_REQUESTED** [review]

File: collect_app/src/test/java/org/odk/collect/android/preferences/source/PreferencesDataStoreTest.kt (hunk 6 above)
Comment (grzesiek2010) [reply collect_app/src/test/java/org/odk/collect/android/preferences/source/PreferencesDataStoreTest.kt]: I tend to agree so I improved it according to what you said.

File: collect_app/src/main/java/org/odk/collect/android/preferences/source/PreferencesDataStore.kt (hunk 7 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/preferences/source/PreferencesDataStore.kt]: If class naming is the last thing maybe we could merge to avoid more conflicts and continue this discussion on slack?

Comment (lognaturel) [reply collect_app/src/main/java/org/odk/collect/android/preferences/source/PreferencesDataStore.kt]: I have a preference for resolving the issue in this PR if possible to try to keep history somewhat easier to navigate. Big renames can be really hard to follow. What are you seeing conflicts with, @grzesiek2010? Always helpful to make conflict concerns concrete so it's easier to evaluate! Looks like some of the renaming needs to happen in master anyway so maybe it's not a big deal in this case.

I think using the word `Settings` for Collect-specific classes is a good idea since that matches the UI. Some ideas:
`PreferencesDataSource` -> `Settings`
`defaultPreferences` -> `settingKeysToDefaults`
`loadDefaultPreferencesIfNotExist` -> `setDefaultForAllSettingsWithoutValues`
`OnPreferenceChangeListener` -> `OnSettingChangeListener`
`SharedPreferencesDataSource` -> `SharedPreferencesSettings`
`PreferencesDataSourceProvider` -> `SettingsProvider` (`getMetaSettings`, `getGeneralSettings`, etc)

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/preferences/source/PreferencesDataStore.kt]: Yeah for me, it's easier to make the changes here as I'm going to have to rebase on top either way. I'm happy with the general change to "Settings".

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/preferences/source/PreferencesDataStore.kt]: Ok renamed those classes/methods.

**Review by seadowg — APPROVED** [review]
